GHA: Grant Collaborator status to contributors who close 2 Issues in past month

zrhoffman opened the pull request:
<!--
Thank you for contributing! Please be sure to read our contribution guidelines: https://github.com/apache/trafficcontrol/blob/master/CONTRIBUTING.md
If this closes or relates to an existing issue, please reference it using one of the following:

Closes: #ISSUE
Related: #ISSUE

If this PR fixes a security vulnerability, DO NOT submit! Instead, contact
the Apache Traffic Control Security Team at security@trafficcontrol.apache.org and follow the
guidelines at https://apache.org/security regarding vulnerability disclosure.
-->

This PR adds the *Assign Triage Role* GitHub Action, which
* Runs on the first day of every month
* Gathers a list of contributors without push access who have had 2 or more Issue-closing Pull Requests merged in the past 31 days
* Creates a Pull Request that summarizes their contributions, explains what ATC [collaborators](https://s.apache.org/asfyamltriage) are, and adds their usernames to `.asf.yaml`.
<!-- **^ Add meaningful description above** --><hr>

## Which Traffic Control components are affected by this PR?
<!-- Please delete all components from this list that are NOT affected by this PR.
Feel free to add the name of a tool or script that is affected but not on the list.
-->
- Automation - GitHub Actions<!-- Please specify which (GitHub Actions, Docker images, Ansible Roles, etc.) -->
- Other - `.asf.yaml`

## What is the best way to verify this PR?
<!-- Please include here ALL the steps necessary to test your PR.
If your PR has tests (and most should), provide the steps needed to run the tests.
If not, please provide step-by-step instructions to test the PR manually and explain why your PR does not need tests. -->
1. 
    1. Push the [`zrhoffman:atc-collaborators`](https://github.com/zrhoffman/trafficcontrol/tree/atc-collaborators) branch to the `master` branch of your own ATC fork
    2. Go to Actions -> *Assign Triage Role*
    3. To the right of the text on the blue bar saying *This workflow has a `workflow_dispatch` event trigger.*, click the gray *Run workflow* dropdown
    4. Set *Count commits starting X days ago* from 31 to around 60 for the purposes of testing, in case no one has had 2 Issue-closing Pull Requests merged within the past 31 days
    5. Click the green *Run workflow* button, then wait for the run of the *Assign Triage Role* workflow to complete (on my fork, it usually takes around 40 seconds).
    6. Still on your fork, go to *Pull requests* and open the Pull Request titled *ATC Collaborators for December 2021*
    7. Look at the generated Pull Request, form opinions

    **Note:** The `#` and `@` symbols are intentionally replaced with similar-looking characters when the workflow runs on forks so that PRs and Users are not unnecessarily mentioned.

2. Run the `assign_triage_role` tests:
   ```shell
   cd .github/actions/assign-triage-role
   python3 -m unittest discover ./tests

## PR submission checklist
- [x] This PR has tests <!-- If not, please delete this text and explain why this PR does not need tests. -->
- [x] This PR has documentation <!-- If not, please delete this text and explain why this PR does not need documentation. -->
- [x] This PR has a CHANGELOG.md entry <!-- A fix for a bug from an ATC release, an improvement, or a new feature should have a changelog entry. -->
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

ocket8888 commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - [#6032](https://github.com/apache/trafficcontrol/issues/6032) Add t3c setting mode 0600 for secure files
 - [#6405](https://github.com/apache/trafficcontrol/issues/6405) Added cache config version to all t3c apps and config file headers
 - Traffic Vault: Added additional flag to TV Riak (Deprecated) Util
+- ATC Collaborators: Added a GitHub Action to grant the GitHub Triage role for a month to contributors who closed 2 Issues in the past month
```
This isn't actually a change to the software that's present in any given version. Collaborators will have permissions on GitHub, and can e.g. triage backport issues for old versions. It's not a change of which people using the software need to be made aware. Which is to say, I don't think this needs a changelog entry.

zrhoffman replied:
Removed changelog entry in 3dd11ef81a

ocket8888 reviewed (commented):
I haven't run this through our Pylint configuration, but I can tell you right now it won't like all of the single-quoted strings and no docstrings anywhere.

One other thing I've noticed is the use of typing on every declaration everywhere - a lot of the time that's unnecessary IMO as the type can be trivially deduced from the assignment. Something like `pr: Optional[PullRequest] = None` makes sense because otherwise I'd just assume that it's `NoneType` based on that assignment. However, `contributor: str = str()` is extraneous, because the trivial assumption that the type is that of the return type of `str` is correct.

One of the better examples of this that I've found is:

```python3
prs_by_contributor: dict[NamedUser, list[(Issue, Issue)]] = dict[NamedUser, list[(Issue, Issue)]]()
```

which, to avoid annoying duplication that makes edits a bit more contrived than they need to be, should probably be
```python3
prs_by_contributor = dict[NamedUser, list[(Issue, Issue)]]()
```
or, even better (because it's something like 4 instructions faster):
```python3
prs_by_contributor: dict[NamedUser, list[(Issue, Issue)]] = {}
```
I wouldn't say, though, that these typings are only extraneous when being assigned a direct, primitive value. Like if I have
```python3
def some_func() -> str:
    """Does some stuff"""
    # ...
```
then it'd be extraneous to have
```python3
my_var: str = some_func()
```
It's still trivial to deduce the type from that assignment. I do think return and argument typings are always necessary, even though technically the return type can be inferred in much the same way, since it enforces usage typing and makes breaking changes to that usage more intentional.

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+	def get_committers(self) -> dict[str, None]:
+		committers: list[str] = sorted([user.login for user in self.repo.get_collaborators() if user.permissions.push])
+		committers_dict: dict[str, None] = {committer: None for committer in committers}
+		return committers_dict
```
Changed to a `set` in d67ddc01e8

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+
+	def get_repo_name(self) -> str:
+		repo_name: str = self.getenv(ENV_GITHUB_REPOSITORY)
+		return repo_name
```
Removed `TriageRoleAssigner.get_repo_name()` in bcbdaf99d9

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+			repo: Repository = self.gh.get_repo(repo_name)
+		except BadCredentialsException:
+			print(f'Credentials from {ENV_GITHUB_TOKEN} were bad.')
+			sys.exit(1)
```
Moved `try`/`except` to `__main__` in 0abb05f55e

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/templates/single_contributor.md:
```diff
@@ -0,0 +1,2 @@
+* @{CONTRIBUTOR_USERNAME} fixed {CONTRIBUTION_COUNT} Issues:
```
Added SPDX license identifier in e45e9b7503

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/templates/single_pr.md:
```diff
@@ -0,0 +1 @@
+    - #{ISSUE_NUMBER}, fixed by #{PR_NUMBER}
```
Added SPDX license identifier in e45e9b7503

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/__main__.py:
```diff
@@ -0,0 +1,31 @@
+	gh = Github(login_or_token=github_token)
+	TriageRoleAssigner(gh).run()
+
+main()
```
Moved logic out of `main()` in 66cdc8529c

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+from github.PaginatedList import PaginatedList
+from github.PullRequest import PullRequest
+from github.Repository import Repository
+from yaml import YAMLError
```
Doing `from yaml import YAMLError` after `import yaml` is an issue? Is there a pylint rule for that?

zrhoffman replied:
In any case, the `YamlError` import was moved to `__main__` in f5207efd64.

ocket8888 replied:
It's not an issue in CPython, dunno if there's a rule for it in Pylint. I'd probably feel better about it if they were at least grouped right next to each other.

zrhoffman replied:
That was the order that PyCharm put them in (ordered `import`, then ordered `from [...] import`). I agree that having them so far apart looks weird, though.

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+					continue
+				pull_request = Issue(self.repo._requester, event.raw_headers, pr_text, completed=True)
+			# Skip unmerged PRs
+			if 'merged_at' not in pull_request.pull_request.raw_data:
```
Moved the check below the `pull_request is None` check in 722895aa60

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+			if author.login in committers:
+				continue
+			if author not in prs_by_contributor:
+				prs_by_contributor[author] = list[(Issue, Issue)]()
```
Using `[]` in db7db0a200

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+		return prs_by_contributor
+
+	def ones_who_meet_threshold(self, prs_by_contributor: dict[NamedUser, list[(Issue, Issue)]]) -> dict[str, list[(Issue, Issue)]]:
+		prs_by_contributor: dict[str, list[(Issue, Issue)]] = {
```
`return`ing the value in d14c51d23b

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+			return True
+		except GithubException as e:
+			message = e.data.get('message')
+			if not re.match(r'No commit found for the ref', message):
```
Checking the response status code instead in 5a117a32ce

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+		with open(os.path.join(os.path.dirname(__file__), PR_TEMPLATE_FILE)) as stream:
+			pr_template = stream.read()
+
+		contrib_list_list: str = str()
```
No longer using `str()` as of f442dad613

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+			pr_list: str = ''
+			for pr, linked_issue in pr_tuples:
+				pr_line = pr_line_template.format(ISSUE_NUMBER=linked_issue.number, PR_NUMBER=pr.number)
+				pr_list += pr_line + '\n'
```
Using `str.join()` in f442dad613

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+			process_label: Label = self.repo.get_label('process')
+			pr.add_to_labels(collaborators_label, process_label)
+		except UnknownObjectException:
+			print('Unable to find a label named "collaborators".')
```
Printing to stderr in f9dbfe7b98

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+
+	def get_repo_owner(self) -> str:
+		repo_name: str = self.getenv(ENV_GITHUB_REPOSITORY_OWNER)
+		return repo_name
```
Made environment variables module-level constants in bcbdaf99d9

zrhoffman commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,279 @@
+
+
+class TriageRoleAssigner:
+	gh: Github
```
Extending `GitHub` class in 1a91d0dea4

zrhoffman commented:
> I haven't run this through our Pylint configuration, but I can tell you right now it won't like all of the single-quoted strings and no docstrings anywhere.

Added docstrings in af30ba1a23, but our Pylint configuration did not complain about single-quoted strings. Switched to double-quoted strings in c961ceb50e, though our Python coding standards should mention it somewhere if we follow it.

* Changed named list and dict declarations to brackets in db7db0a200
* Removed unnecessary typehints in 1f26c4959a

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,342 @@
+from github.Repository import Repository
+from yaml import YAMLError
+
+from assign_triage_role.constants import GH_TIMELINE_EVENT_TYPE_CROSS_REFERENCE, ASF_YAML_FILE, \
+	APACHE_LICENSE_YAML, GIT_AUTHOR_EMAIL_TEMPLATE, SINGLE_PR_TEMPLATE_FILE, \
+	SINGLE_CONTRIBUTOR_TEMPLATE_FILE, PR_TEMPLATE_FILE, EMPTY_CONTRIB_LIST_LIST, \
+	EMPTY_LIST_OF_CONTRIBUTORS, CONGRATS, EXPIRE, GITHUB_REPOSITORY, GIT_AUTHOR_NAME, \
+	GITHUB_REPOSITORY_OWNER, MINIMUM_COMMITS, SINCE_DAYS_AGO, GITHUB_REF_NAME
```
If you want to avoid using backslashes on every line, you can use parenthesis to group the imports into a single multiline statement. But you don't have to.

zrhoffman replied:
Parenthesizing imports in 9601a60e5d

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,342 @@
+	"""
+	Triage Role Assigner
+	"""
+	github: Github
```
unused class member

zrhoffman replied:
Removed class member in 110291d6f5

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,342 @@
+
+	def get_committers(self) -> set[str]:
+		"""
+		Gets a dict whose keys are the usernames of committers
```
it's actually a set

zrhoffman replied:
Documented as a set in 8cc43ea274

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,342 @@
+			try:
+				asf_yaml: dict[str, dict] = yaml.safe_load(stream)
+			except YAMLError as e:
+				print(f"Could not load YAML file {ASF_YAML_FILE}: {e}")
```
should this be printing to stderr?

zrhoffman replied:
Printing to stderr in f5207efd64

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/triage_role_assigner.py:
```diff
@@ -0,0 +1,342 @@
+				asf_yaml: dict[str, dict] = yaml.safe_load(stream)
+			except YAMLError as e:
+				print(f"Could not load YAML file {ASF_YAML_FILE}: {e}")
+				sys.exit(1)
```
should this be handled in __main__ like the other exception?

zrhoffman replied:
Handling `YAMLError` in `__main__` in f5207efd64

ocket8888 commented on .github/actions/assign-triage-role/assign_triage_role/templates/pr_template.md:
```diff
@@ -0,0 +1,66 @@
+If this PR fixes a security vulnerability, DO NOT submit! Instead, contact
+the Apache Traffic Control Security Team at security@trafficcontrol.apache.org and follow the
+guidelines at https://apache.org/security regarding vulnerability disclosure.
+-->
```
If I ever have to edit this template, the first thing I'm going to do is get rid of all these comments meant for human authors on GitHub. They aren't hurting anything, but I'm just saying it'd increase the signal-to-noise ratio in this file.

zrhoffman replied:
Removed human-intended comments in acc658fce1

ocket8888 commented:
> our Pylint configuration did not complain about single-quoted strings

My mistake. The ESLint configuration(s) we have care, I think, so I probably just got those confused. If the Pylint config doesn't have a rule set for it, then it's not a rule. It's a bit late to correct myself, unfortunately, since you already changed them all.
